### PR TITLE
[clang][headers] Including stddef.h always redefines NULL

### PR DESCRIPTION
**Pull request by ian-twilightcoder**
stddef.h always includes __stddef_null.h. This is fine in modules because it's not possible to re-include the pcm, and it's necessary to export the _Builtin_stddef.null submodule. However, without modules it causes NULL to always get redefined which disrupts some C++ code. Rework the inclusion of __stddef_null.h so that with not building with modules it's only included if __need_NULL is set by the includer, or it's the first time stddef.h is being included.

### CODE REVIEW

File: clang/lib/Headers/stddef.h
Comment (zygoloid) [inline clang/lib/Headers/stddef.h:51]: I'm not sure that the second condition here makes any difference: I'd think the only way we could get here with `__building_module(_Builtin_stddef)` evaluating to `true` is if one of the modular headers in that module (`__stddef_*.h`) were to recursively include `stddef.h`, which it seems they shouldn't be doing! Also, given that we have the actual definition of the `NULL` macro in a non-textual module of its own these days, it doesn't seem like the problem described below would actually happen.

Can we drop the `__building_module` check here?

I also wonder if we can move the `!defined(__STDDEF_H)` condition up to the `#if` on line 38 -- if we've already had a full include of the entirety of `<stddef.h>`, we shouldn't `__need` anything, whether it's a macro or something else.

Comment (ian-twilightcoder) [reply clang/lib/Headers/stddef.h:51]: We do still need everything when we're building with modules for the reasons outlined in the comment, but I think you're right that `__building_module(_Builtin_stddef)` doesn't do anything and that needs to be `__has_feature(modules)`

**Review by AaronBallman — COMMENTED** [review]
These changes should come with a release note so users know about the fix.

**Comment by ian-twilightcoder** [comment]
> These changes should come with a release note so users know about the fix.

Does it need a release note? It was introduced in 19 with https://github.com/llvm/llvm-project/pull/90676, and this will fix it in 19 too.

File: clang/lib/Headers/stdarg.h
Comment (ian-twilightcoder) [reply clang/lib/Headers/stdarg.h:28]: clang/test/Headers/stddefneeds.cpp already tests this, but let me add one in clang/test/Modules too

**Comment by AaronBallman** [comment]
> > These changes should come with a release note so users know about the fix.
> 
> Does it need a release note? It was introduced in 19 with #90676, and this will fix it in 19 too.

Ah nope, no need for a release note then!

**Review by AaronBallman — COMMENTED** [review]
With additional test coverage, the changes LGTM but I'd appreciate it if @zygoloid or @ChuanqiXu9 could validate the modules logic.

File: clang/lib/Headers/stdarg.h
Comment (ian-twilightcoder) [reply clang/lib/Headers/stdarg.h:28]: clang/test/Headers/stddefneeds.cpp verifies that NULL does and doesn't get redeclared when stddef.h gets included multiple times without modules, and clang/test/Modules/stddef.cpp verifies that NULL always gets declared when stddef.h is transitively included by a submodule.

File: clang/lib/Headers/stddef.h
Comment (ian-twilightcoder) [reply clang/lib/Headers/stddef.h]: clang/test/Modules/stddef.cpp shows how even if we've already had a full include of the entirety of `<stddef.h>`, we still `__need` everything. It also shows that you're right and `__building_module(_Builtin_stddef)` doesn't work, but `__has_feature(modules)` does.

File: clang/lib/Headers/stddef.h
Comment (zygoloid) [inline clang/lib/Headers/stddef.h:68]: This change now looks like a (fairly trivial, but still) regression for local submodule visibility mode. In particular:
```c++
#include <stddef.h>
#undef NULL
#include <stddef.h>
```
is supposed to leave `NULL` undefined, but won't do so any more.

Perhaps we should introduce a feature test macro for "no local submodule visibility" mode ("name leakage between submodules" mode?) so that workarounds for the infelicities of that mode remain scoped to it?

Comment (ian-twilightcoder) [reply clang/lib/Headers/stddef.h:68]: I'm pretty sure that will leave `NULL` undefined even in LSV because the second include won't re-bring in any of the pcms but I can add that to the test.

Comment (ian-twilightcoder) [reply clang/lib/Headers/stddef.h:68]: (I know that because I was trying to make clang/test/Headers/stddefneeds.cpp work in modules and it simply isn't possible according to some of the reviewers in https://reviews.llvm.org/D159064)

Comment (ian-twilightcoder) [reply clang/lib/Headers/stddef.h:68]: Added a "anotherPointer" test to clang/test/Modules/stddef.cpp to illustrate.

Comment (zygoloid) [reply clang/lib/Headers/stddef.h:68]: OK, I see: we want a plain include of `<stddef.h>` to define `NULL` only the first time. In non-modules builds, we get this by checking for `__STDDEF_H`, and in modules builds we get this because `__stddef_null.h` is a modular header. (And checking for `__STDDEF_H` isn't sufficient for a modules but non-LSV build because it might leak from a different submodule.) So it doesn't matter whether we define `__need_NULL` for a modules LSV build; it does the same thing either way.

It looks like we would give different behavior in a modules build for something like:
```
#define __need_NULL
#include <stddef.h>
#undef NULL

#include <stddef.h>
```
... which normally results in `NULL` being defined, but with any flavor of modules, results in `NULL` *not* being defined -- because we build `__stddef_null.h` as a modular header, even though it's really not (it's supposed to redefine `NULL` each time it's included). But that's a pre-existing issue.

In any case, I think I'm convinced this change is correct. Thanks for bearing with me :)

Comment (ian-twilightcoder) [reply clang/lib/Headers/stddef.h:68]: > we build __stddef_null.h as a modular header, even though it's really not (it's supposed to redefine NULL each time it's included)

But it _has_ to be a modular header because otherwise all the includers would declare `NULL` in their pcm so we'd have A.NULL, B.NULL, etc. Though I don't think you generally use `NULL` in Swift, if you somehow did (like `func doSomething(pointer: <something>? = NULL)` maybe?) that would cause problems.

Comment (zygoloid) [reply clang/lib/Headers/stddef.h:68]: As a general principle, it seems to me that getting C and C++ semantics right should be the primary concern here, and we shouldn't regress correctness in C or C++ to improve the behavior of Swift interop. Though as it happens, in this case the modules behavior does seem like a reasonable choice for C and C++ -- and really the issue is just that we have a divergence in behavior between modules and non-modules compilations.

I wonder if we could change our behavior for `__need_*` in non-modules builds to match the modules behavior: specifically that you only get each piece once, the first time you ask for it, so if you ask for just `NULL` then ask for everything, you don't get `NULL` the second time. If we think that's good enough for modules, then it seems like it should be good enough in general, and means there's less that can go wrong when turning on / off modules.

Comment (ian-twilightcoder) [reply clang/lib/Headers/stddef.h:68]: I agree that would be the more predictable behavior; I'm generally opposed with redefining `NULL` or relying on clang to participate in redefinition schemes. But I don't know how much changing that part would break clients, so I'd rather do it in a separate change since I'm _just_ trying to fix the regression building gcc with clang in this fix.

Comment (zygoloid) [reply clang/lib/Headers/stddef.h:68]: Yeah, sorry, I should have been clear: I still think this change is correct as-is and I'm not suggesting making changes in this pull request.

**Comment by ldionne** [comment]
@AaronBallman Should this be added to the 19.x milestone and cherry-picked?

**Comment by AaronBallman** [comment]
> @AaronBallman Should this be added to the 19.x milestone and cherry-picked?

I think it should be

**Comment by ian-twilightcoder** [comment]
> > @AaronBallman Should this be added to the 19.x milestone and cherry-picked?
> 
> I think it should be

Nuts, should've merged it yesterday.

**Comment by ldionne** [comment]
@ian-twilightcoder Cherry-picking is easy. Once this is merged, follow the guidelines here to get it cherry-picked: https://llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches

Basically you only need to add a comment to this PR like `/cherry-pick COMMIT-SHA` and everything else pretty much happens automatically.

**Comment by ian-twilightcoder** [comment]
The Driver/linker-wrapper-passes.c test failure doesn't look related, that test doesn't use stddef.h or anything else. I keep trying to rebase to make it go away but it's stubborn.

**Review by AaronBallman — APPROVED** [review]
> The Driver/linker-wrapper-passes.c test failure doesn't look related, that test doesn't use stddef.h or anything else. I keep trying to rebase to make it go away but it's stubborn.

I looked at the failures as well and I think they're unrelated.

This LGTM and based on the last comment from @zygoloid, I think this is ready to land (and be cherry-picked to the 19.x release).

Thank you for the fix!

**Comment by ian-twilightcoder** [comment]
> @ian-twilightcoder Cherry-picking is easy. Once this is merged, follow the guidelines here to get it cherry-picked: https://llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches
> 
> Basically you only need to add a comment to this PR like `/cherry-pick COMMIT-SHA` and everything else pretty much happens automatically.

Oh I thought you had to make a GitHub issue and get a branch manager to merge it for you. Ok let me give it a try 🤞

**Comment by ldionne** [comment]
@ian-twilightcoder You can create an issue if you want, but you can also say /cherry-pick <COMMIT> directly from the PR as long as you add the release milestone to it.

That being said, the cherry-pick failed so you will need to create a PR manually.

**Pull request merged**
